lib/charms/observability_libs/v0/kubernetes_compute_resources_patch.py

```diff
@@ -468,51 +489,67 @@ def _on_config_changed(self, _):
         self._patch()
 
     def _patch(self) -> None:
-        """Patch the Kubernetes resources created by Juju to limit cpu or mem."""
+        """Patch the Kubernetes resources created by Juju to limit cpu or mem.
+
+        This method will keep on retrying to patch the kubernetes resource for a default duration of 3 minutes,
+        if the patching failure is due to a recoverable error (e.g: Network Latency).
+        This will put the charm in a `Maintenance` state until the patch is either successful or all retries have been consumed,
+        which in that case, will raise the same exception as the original exception that caused it and will put the charm in error state.
+        If the failure is not recoverable (e.g: wrong spec provided), an exception same as the original exception that caused it will be raised and will put the charm in error state.
+        """
         try:
             resource_reqs = self.resource_reqs_func()
             limits = resource_reqs.limits
             requests = resource_reqs.requests
         except ValueError as e:
             msg = f"Failed obtaining resource limit spec: {e}"
-            logger.error(msg)
-            self.on.patch_failed.emit(message=msg)
-            return
+            logger.exception(msg)
+            raise ValueError(msg) from e
```

Comment:
`_patch` is [called on config-changed](https://github.com/canonical/observability-libs/blob/a6ac58eb30286d34b269a1c585e5c8cc894ae373/lib/charms/observability_libs/v0/kubernetes_compute_resources_patch.py#L488), which means there is no obvious place to place a try-except guard in the charm: a charm instantiates the resource patch obj, which calls and observe, which goes into a queue, and only later, when ops processes the events, we'll get an exception.

I.e. it seems that with this design charm authors would need to wrap the call to `ops.main` with try-except, or intentionally/knowingly never try to catch it. I.e. semantically this is similar to [SystemExit](https://docs.python.org/3/library/exceptions.html#SystemExit). Are we sure we want to introduce this kind of behavior?

Reply:
The intention is not to catch it. Its to let it go to error state, as if you ever reach that point, there's no point of catching the error. The charm can do nothing and you'll need to manually fix sth.

Reply:
I'm afraid users would not like it:
- Event handlers are processed by the order they were observed. What if there are other config-changed concerns that are now not being executed because our `observe` call came first? Do we really want an error state just because a resource limit failed to apply? Not sure this justifies a total outage.
- We support refresh events so the above may happen for other events, event custom events, that are passed in.

I think a SystemExit-like side effect from a charm lib is a big deviation from what most expect from charm libs. For example, one convention that we imposed on ourselves is that charmlibs shall not write unit status.

Reply:
Refactored logic to not have any error states. Updated PR description.